wgpu-hal/src/gles/wgl.rs

```diff
@@ -0,0 +1,750 @@
+            String::from("unable to get executable instance"),
+            Error::last_os_error(),
+        ));
+    }
+
+    // Use the address of `UNIQUE` as part of the window class name to ensure different
+    // `wgpu` versions use different names.
+    static UNIQUE: Mutex<u8> = Mutex::new(0);
+    let class_addr: *const _ = &UNIQUE;
+    let name = format!("wgpu Device Class {:x}\0", class_addr as usize);
+    let name = CString::from_vec_with_nul(name.into_bytes()).unwrap();
+
+    unsafe extern "system" fn wnd_proc(
+        window: HWND,
+        msg: UINT,
+        wparam: WPARAM,
+        lparam: LPARAM,
+    ) -> LRESULT {
+        unsafe { DefWindowProcA(window, msg, wparam, lparam) }
+    }
+
+    let window_class = WNDCLASSEXA {
+        cbSize: mem::size_of::<WNDCLASSEXA>() as u32,
+        style: CS_OWNDC,
+        lpfnWndProc: Some(wnd_proc),
+        cbClsExtra: 0,
+        cbWndExtra: 0,
+        hInstance: instance,
+        hIcon: ptr::null_mut(),
+        hCursor: ptr::null_mut(),
+        hbrBackground: ptr::null_mut(),
+        lpszMenuName: ptr::null_mut(),
+        lpszClassName: name.as_ptr(),
+        hIconSm: ptr::null_mut(),
+    };
+
+    let atom = unsafe { RegisterClassExA(&window_class) };
+
+    if atom == 0 {
+        return Err(crate::InstanceError::with_source(
+            String::from("unable to register window class"),
+            Error::last_os_error(),
+        ));
+    }
+
+    let window = unsafe {
+        CreateWindowExA(
+            0,
+            name.as_ptr(),
+            name.as_ptr(),
+            0,
+            0,
+            0,
+            1,
+            1,
+            ptr::null_mut(),
+            ptr::null_mut(),
+            instance,
+            ptr::null_mut(),
+        )
+    };
+    if window.is_null() {
+        return Err(crate::InstanceError::with_source(
+            String::from("unable to create hidden instance window"),
+            Error::last_os_error(),
+        ));
+    }
+    let dc = unsafe { GetDC(window) };
+    if dc.is_null() {
+        return Err(crate::InstanceError::with_source(
+            String::from("unable to create memory device"),
+            Error::last_os_error(),
+        ));
+    }
+    unsafe { setup_pixel_format(dc)? };
+
+    // We intentionally leak the window class, window and device context handle to avoid
+    // spawning a thread to destroy them. We cannot use `DestroyWindow` and `ReleaseDC` on
+    // different threads.
+
+    Ok(dc)
+}
+
+fn get_global_device_context() -> Result<HDC, crate::InstanceError> {
+    #[derive(Clone, Copy)]
+    struct SendDc(HDC);
+    unsafe impl Sync for SendDc {}
+    unsafe impl Send for SendDc {}
+
+    static GLOBAL: Mutex<Option<Result<SendDc, crate::InstanceError>>> = Mutex::new(None);
+    let mut guard = GLOBAL.lock();
+    if guard.is_none() {
+        *guard = Some(create_global_device_context().map(SendDc));
+    }
+    guard.clone().unwrap().map(|dc| dc.0)
+}
```

Comment:
Isn't this basically just a `once_cell::Lazy<Result<...>>`?

Reply:
Probably. I didn't want to add a dependency.

Reply:
`once_cell` is fine to add due to its ubiquity